### PR TITLE
add shellcheck lint

### LINKED ISSUE #24614: Add a `shellcheck` based pre-submit
[shellcheck](https://github.com/koalaman/shellcheck) is a bash script linter which could help catch common problems in our (numerous) bash scripts. I know we'd like to reduce our use of bash scripts, but until then I think being stricter about script quality would be helpful.

/cc @zmerlynn

### PR DESCRIPTION
**Pull request by BenTheElder**
**What this PR does / why we need it**: Adds a [`shellcheck`](https://github.com/koalaman/shellcheck) lint to find errors in shell scripts, initially marks all failing scripts as known failures. We can incrementally fix these / disable unnecessary lint error codes going forward.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #24614

**Special notes for your reviewer**: 
- ~we'll still need to add an actual presubmit job using this via PR to [test-infra](https://github.com/kubernetes/test-infra)~ we're adding it to the existing job
- ~this is still WIP because I want to:~
  - ~sanity check by fixing some failures and removing a file from the known failures~
  - ~discuss if there are any other error codes we should ignore~ can be done in follow-ups
  - ~figure out how we want to version / locally install `shellcheck`~ we're now using the official docker image

**Release note**:

```release-note
NONE
```

/sig testing

### CODE REVIEW

**Comment by BenTheElder** [comment]
/assign @cblecker @ixdy
Probably the biggest issue is how we want to have users install shellcheck / how we want to version it in CI (do we want to fetch binaries ourselves?)

The other problem is that it's a little slow. There's possibly some optimizations to be had.

**Comment by dims** [comment]
Suggestion : this is GPLv3 code in https://github.com/koalaman/shellcheck and looks like it's good news that they have published container images https://hub.docker.com/r/koalaman/shellcheck/tags/ - can we please use those?

**Comment by dims** [comment]
we can always optimize it to nail the shellcheck version (and check if there is a local copy and run that instead of needing docker)

**Comment by BenTheElder** [comment]
SGTM, we can probably just run the presubmit in this image too using the podutils. I'll take a look at that soon.

**Comment by BenTheElder** [comment]
/hold
We won't want to merge this prior to the release anyhow. I will re-work this to support the shellcheck docker image when we're a little bit closer to the next cycle.

File: hack/verify-shellcheck.sh [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+cd "${KUBE_ROOT}"
+
+# find all shell scripts excluding ./_* and ./vendor*
+all_shell_scripts=()
+while IFS=$'\n' read -r script;
+  do all_shell_scripts+=("$script");
+done < <(find . -name "*.sh" \
+  -not \( \
+    -path ./_\*      -o \
+    -path ./vendor\*    \
```
[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/verify-shellcheck.sh]: Good point, I'll take a stab at fixing this later today along with supporting running in lints in the docker image.

**Comment by BenTheElder** [comment]
Will try to come back to this in the next cycle, I think this is still important but I haven't had the time to sort out how we wrap the upstream docker image instead.

**Comment by BenTheElder** [comment]
@dims on my laptop the current iteration takes about 35s to run with shellcheck on the host, but a naive swap out of the local binary for running the docker image it is fantastically slow 😞 

will look at optimizing next ...

**Comment by BenTheElder** [comment]
re-using the container brings it to about 2 minutes which seems reasonable enough.

File: hack/verify-shellcheck.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply hack/verify-shellcheck.sh]: this should now ignore git and respect gitignore

**Comment by BenTheElder** [comment]
the last remaining question I have: do we make this part of the default verify suite, or add a new job per my original comment?

**Comment by dims** [comment]
if it's 2-3 mins, am ok to add it to the existing job @BenTheElder

File: hack/verify-shellcheck.sh [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    } >&2
+    exit 1
+fi
+
+# lint each script, tracking failures
+errors=()
+not_failing=()
+for f in "${all_shell_scripts[@]}"; do
+  set +o errexit
+  failedLint=$(docker exec -t "${SHELLCHECK_CONTAINER}" shellcheck --exclude="${SHELLCHECK_DISABLED}" "${f}")
```
Comment (BenTheElder) [inline hack/verify-shellcheck.sh]: for some reason the paths are not working right in CI... investigating.

Comment (BenTheElder) [reply hack/verify-shellcheck.sh]: this was a CI issue, we've updated the verify job

**Comment by BenTheElder** [comment]
@dims @ixdy @cblecker all good to go now, passed in `1m52.414s` https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/68438/pull-kubernetes-verify/118705

**Pull request merged**